### PR TITLE
Implement `RetryProgress` for limiting retry attempts in stages

### PR DESCRIPTION
**Pull request by addisoncrump**
Addresses a scenario where comparison log tracing never completes as it gets stuck in a timeout loop.

### CODE REVIEW

**Comment by addisoncrump** [comment]
Okay, small hiccup: we need the implementation of NeverResume to be generic over the type which it is used by, but the only real way to do this I can find is using the nightmare fuel syntax in the last commit.

We can instead make a macro which defines NeverResume uniquely to avoid this, but... gross.

**Comment by addisoncrump** [comment]
Resolved the issues above by simply thinking.

File: libafl/src/stages/tracing.rs [hunk 1]
```diff
@@ -182,7 +212,16 @@ where
...
         Self {
+            initial_tries: 10,
             phantom: PhantomData,
         }
     }
+
+    /// Apply a certain amount of tries that this stage will attempt to trace the input before
+    /// giving up and not processing the input again.
+    #[must_use]
+    pub fn with_tries(mut self, initial_tries: usize) -> Self {
```
Comment (domenukk) [inline libafl/src/stages/tracing.rs:223]: _re_tries?

File: libafl/src/stages/concolic.rs [hunk 2]
```diff
@@ -39,10 +39,10 @@ where
     E: UsesState<State = TE::State>,
     EM: UsesState<State = TE::State>,
     TE: Executor<EM, Z> + HasObservers,
-    TE::State: HasExecutions + HasCorpus,
+    TE::State: HasExecutions + HasCorpus + HasNamedMetadata,
     Z: UsesState<State = TE::State>,
 {
-    type Progress = (); // stage cannot be resumed
+    type Progress = (); // stage internally handles progress
```
Comment (domenukk) [inline libafl/src/stages/concolic.rs:45]: `type HandlesProgressInternally = ()`?

File: libafl/src/stages/concolic.rs [hunk 3]
```diff
@@ -58,6 +58,17 @@ where
             ));
         };
 
+        <TracingStage<EM, TE, Z> as Stage<E, EM, Z>>::Progress::initialize_progress(
```
Comment (domenukk) [inline libafl/src/stages/concolic.rs:61]: W H A T

I'm surprised that's even legal rust, is there any way to write this in a simpler way?

File: libafl/src/stages/mod.rs [hunk 4]
```diff
@@ -541,38 +548,128 @@ pub mod pybind {
...
+                    skipped: HashSet::new(),
+                },
+                core::any::type_name_of_val(stage),
+            );
+        }
+        Ok(())
+    }
+
+    fn clear_progress(state: &mut S, stage: &ST) -> Result<(), Error> {
+        let metadata = state.named_metadata_mut::<Self>(core::any::type_name_of_val(stage))?;
```
Comment (domenukk) [inline libafl/src/stages/mod.rs:632]: I think we should look up the the metadata once in the stage, every hashmap lookup is pretty expensive

Comment (domenukk) [reply libafl/src/stages/mod.rs:632]: Or better yet, cache it locally in the stage instance? No need to hit the map every round if we didn't restart (`Rc<RefCell<>>` that guy into an Option or something)

**Review by domenukk — APPROVED** [review]

File: libafl/src/stages/mod.rs [hunk 5]
```diff
@@ -15,10 +15,12 @@ pub use concolic::SimpleConcolicMutationalStage;
 #[cfg(feature = "std")]
 pub use dump::*;
 pub use generalization::GeneralizationStage;
```
Comment (domenukk) [inline libafl/src/stages/mod.rs:17]: GeneralizationStage will also need that restarter, right?

File: libafl/src/stages/tracing.rs (hunk 1 above)
Comment (addisoncrump) [reply libafl/src/stages/tracing.rs:223]: `with_retries(1)` would suggest that the input is run twice, not once like current behaviour.

File: libafl/src/stages/concolic.rs (hunk 2 above)
Comment (addisoncrump) [reply libafl/src/stages/concolic.rs:45]: Nice, I'll do this one for now but this is probably worthwhile to add elsewhere as well.

File: libafl/src/stages/concolic.rs (hunk 3 above)
Comment (addisoncrump) [reply libafl/src/stages/concolic.rs:61]: Bit of an inelegance that Rust has... sometimes it needs the specific type that an associated type is being pulled from but here it should be deducible. I'll find a better way to write this.

File: libafl/src/stages/mod.rs (hunk 4 above)
Comment (addisoncrump) [reply libafl/src/stages/mod.rs:632]: If we store this anywhere but in the state, it won't be recovered on resume, so we must eat the lookup. However, given the relative cost of tracing and that this happens once per stage initialisation (either the first time the stage is called for this fuzz cycle or during resume), I don't think this is going to be a perf issue.

File: libafl/src/stages/mod.rs (hunk 5 above)
Comment (addisoncrump) [reply libafl/src/stages/mod.rs:17]: Resolved offline.

File: libafl/src/stages/concolic.rs (hunk 3 above)
Comment (addisoncrump) [reply libafl/src/stages/concolic.rs]: Resolved the same way as above.

**Comment by addisoncrump** [comment]
Gah, new clippy lints. I'll just fix those now...

File: libafl/src/stages/mod.rs (hunk 4 above)
Comment (domenukk) [reply libafl/src/stages/mod.rs:632]: No we can look it up once if it's `None` (so after restart), else we just reuse the pre-looked-up one

Comment (addisoncrump) [reply libafl/src/stages/mod.rs:632]: But this doesn't change how many lookups we do, no? We end up requesting the progress the same amount of times in the end.

Comment (domenukk) [reply libafl/src/stages/mod.rs:632]: but ok maybe not necessary

File: libafl/src/stages/tracing.rs (hunk 1 above)
Comment (addisoncrump) [reply libafl/src/stages/tracing.rs:246]: So, you would prefer that the value represents the number of retries? In other words, 0 retries = 1 try?

Comment (addisoncrump) [reply libafl/src/stages/tracing.rs]: I've just pushed this change.

File: libafl/src/stages/tracing.rs [hunk 6]
```diff
@@ -83,15 +77,63 @@ where
...
+        self.initial_tries
+    }
+}
+
 impl<EM, TE, Z> TracingStage<EM, TE, Z> {
     /// Creates a new default stage
     pub fn new(tracer_executor: TE) -> Self {
         Self {
             tracer_executor,
+            initial_tries: 10,
```
Comment (domenukk) [inline libafl/src/stages/tracing.rs:124]: Should this be named `max_retries`?

File: libafl/src/stages/mod.rs [hunk 7]
```diff
@@ -541,38 +548,128 @@ pub mod pybind {
...
+    fn progress<'a>(state: &'a S, stage: &ST) -> Result<&'a Self, Error> {
+        state.named_metadata::<Self>(core::any::type_name_of_val(stage))
+    }
+
+    fn progress_mut<'a>(state: &'a mut S, stage: &ST) -> Result<&'a mut Self, Error> {
+        state.named_metadata_mut::<Self>(core::any::type_name_of_val(stage))
+    }
+}
+
+impl LimitedTriesProgress {
```
[the remark this answers is not in the document]
Comment (domenukk) [reply libafl/src/stages/mod.rs:646]: Or `RetryingProgress`

**Pull request merged**